File: src/main/java/org/jetbrains/kotlin/resolve/NetBeansAnalyzerFacadeForJVM.java
/*******************************************************************************
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *******************************************************************************/
package org.jetbrains.kotlin.resolve;

import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.intellij.openapi.project.Project;
import com.intellij.psi.search.GlobalSearchScope;
import java.util.Collection;
import java.util.List;
import java.util.Set;
import kotlin.Pair;
import org.jetbrains.kotlin.utils.ProjectUtils;
import org.jetbrains.kotlin.analyzer.AnalysisResult;
import org.jetbrains.kotlin.cli.jvm.compiler.CliLightClassGenerationSupport;
import org.jetbrains.kotlin.config.LanguageVersion;
import org.jetbrains.kotlin.container.StorageComponentContainer;
import org.jetbrains.kotlin.context.MutableModuleContext;
import org.jetbrains.kotlin.frontend.java.di.ContainerForTopDownAnalyzerForJvm;
import org.jetbrains.kotlin.incremental.components.LookupTracker;
import org.jetbrains.kotlin.load.java.lazy.LazyJavaPackageFragmentProvider;
import org.jetbrains.kotlin.psi.KtFile;
import org.jetbrains.kotlin.resolve.TopDownAnalysisMode;
import org.jetbrains.kotlin.resolve.jvm.TopDownAnalyzerFacadeForJVM;
import org.jetbrains.kotlin.resolve.lazy.declarations.FileBasedDeclarationProviderFactory;
import org.jetbrains.kotlin.utils.KotlinFrontEndException;


/**
 *
 * @author Александр
 */
public class NetBeansAnalyzerFacadeForJVM {
    
    public static NetBeansAnalyzerFacadeForJVM INSTANCE = new NetBeansAnalyzerFacadeForJVM();
    
    private NetBeansAnalyzerFacadeForJVM(){}
    
    public AnalysisResultWithProvider analyzeFilesWithJavaIntegration(org.netbeans.api.project.Project kotlinProject,
            Project project, Collection<KtFile> filesToAnalyze){
        
        Set<KtFile> filesSet = Sets.newLinkedHashSet(filesToAnalyze);
        if (filesSet.size() != filesToAnalyze.size()){
            System.out.println("Files have duplicates");
        }
        
        Set<KtFile> allFiles = Sets.newLinkedHashSet(filesSet);
        
        Set<String> addedFiles = Sets.newLinkedHashSet();
        for (KtFile file : filesSet){
            if (getPath(file) != null){
                addedFiles.add(getPath(file));
            }
        }
        
        for (KtFile file : ProjectUtils.getSourceFilesWithDependencies(kotlinProject)){
            if (!addedFiles.contains(getPath(file))){
                allFiles.add(file);
            }
        }
        
        MutableModuleContext moduleContext = TopDownAnalyzerFacadeForJVM.createContextWithSealedModule(
                project, project.getName());
        FileBasedDeclarationProviderFactory providerFactory =
                new FileBasedDeclarationProviderFactory(moduleContext.getStorageManager(), allFiles);
        CliLightClassGenerationSupport.CliBindingTrace trace = 
                new CliLightClassGenerationSupport.CliBindingTrace();
                
        Pair<ContainerForTopDownAnalyzerForJvm, StorageComponentContainer> containerAndProvider =
                Injection.createContainerForTopDownAnalyzerForJvm(moduleContext, trace, providerFactory, 
                        GlobalSearchScope.allScope(project), kotlinProject, LookupTracker.Companion.getDO_NOTHING(), 
                        new KotlinPackagePartProvider(kotlinProject), LanguageVersion.LATEST);
        
        ContainerForTopDownAnalyzerForJvm container = containerAndProvider.getFirst();
        List<LazyJavaPackageFragmentProvider> additionalProviders = 
                Lists.newArrayList(container.getJavaDescriptorResolver().getPackageFragmentProvider());
        
        try {
            container.getLazyTopDownAnalyzerForTopLevel().analyzeFiles(TopDownAnalysisMode.TopLevelDeclarations, 
                   filesSet, additionalProviders);
        } catch (KotlinFrontEndException e){
        }
        
        return new AnalysisResultWithProvider(
            AnalysisResult.success(trace.getBindingContext(), moduleContext.getModule()), 
                containerAndProvider.getSecond());
    }
    
    private String getPath(KtFile ktFile){
        return ktFile.getVirtualFile().getPath();
    }
    
}

File: src/main/java/org/jetbrains/kotlin/resolve/NetBeansAnalyzerFacadeForJVM.kt
/*******************************************************************************
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *******************************************************************************/
package org.jetbrains.kotlin.resolve

import com.intellij.psi.search.GlobalSearchScope
import org.jetbrains.kotlin.analyzer.AnalysisResult
import org.jetbrains.kotlin.builtins.JvmBuiltInsPackageFragmentProvider
import org.jetbrains.kotlin.cli.jvm.compiler.CliLightClassGenerationSupport
import org.jetbrains.kotlin.config.LanguageVersion
import org.jetbrains.kotlin.container.ComponentProvider
import org.jetbrains.kotlin.config.CommonConfigurationKeys
import org.jetbrains.kotlin.config.CompilerConfiguration
import org.jetbrains.kotlin.config.LanguageVersionSettingsImpl
import org.jetbrains.kotlin.context.MutableModuleContext
import org.jetbrains.kotlin.context.ProjectContext
import org.jetbrains.kotlin.util.KotlinFrontEndException
import org.jetbrains.kotlin.frontend.java.di.initJvmBuiltInsForTopDownAnalysis
import org.jetbrains.kotlin.incremental.components.LookupTracker
import org.jetbrains.kotlin.context.ContextForNewModule
import org.jetbrains.kotlin.descriptors.PackageFragmentProvider
import org.jetbrains.kotlin.name.Name
import org.jetbrains.kotlin.platform.JvmBuiltIns
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.kotlin.resolve.KotlinPackagePartProvider
import org.jetbrains.kotlin.resolve.TopDownAnalysisMode
import org.jetbrains.kotlin.resolve.jvm.JavaDescriptorResolver
import org.jetbrains.kotlin.resolve.jvm.TopDownAnalyzerFacadeForJVM
import org.jetbrains.kotlin.resolve.jvm.TopDownAnalyzerFacadeForJVM.SourceOrBinaryModuleClassResolver
import org.jetbrains.kotlin.resolve.jvm.extensions.PackageFragmentProviderExtension
import org.jetbrains.kotlin.resolve.lazy.declarations.FileBasedDeclarationProviderFactory
import org.jetbrains.kotlin.resolve.lazy.KotlinCodeAnalyzer
import org.jetbrains.kotlin.model.KotlinEnvironment
import org.jetbrains.kotlin.descriptors.impl.CompositePackageFragmentProvider
import org.jetbrains.kotlin.descriptors.impl.ModuleDependenciesImpl
import java.util.LinkedHashSet
import org.jetbrains.kotlin.utils.ProjectUtils
import org.netbeans.api.project.Project as NBProject
import com.intellij.openapi.project.Project

object NetBeansAnalyzerFacadeForJVM {

    public fun analyzeFilesWithJavaIntegration(
            kotlinProject: NBProject,
            project : Project,
            filesToAnalyze: Collection<KtFile>): AnalysisResultWithProvider {
        val filesSet = filesToAnalyze.toSet()
        
        val allFiles = LinkedHashSet<KtFile>(filesSet)
        val addedFiles = filesSet.map { getPath(it) }.filterNotNull().toSet()
        ProjectUtils.getSourceFilesWithDependencies(kotlinProject).filterNotTo(allFiles) {
            getPath(it) in addedFiles
        }
        
        val environment = KotlinEnvironment.getEnvironment(kotlinProject)
        
        val moduleContext = createModuleContext(project, 
                environment.configuration,
                true)
        val storageManager = moduleContext.storageManager
        val module = moduleContext.module
        
        val providerFactory = FileBasedDeclarationProviderFactory(storageManager, allFiles)
        val trace = CliLightClassGenerationSupport.CliBindingTrace()
        
        val sourceScope = TopDownAnalyzerFacadeForJVM.newModuleSearchScope(project, filesToAnalyze)
        val moduleClassResolver = SourceOrBinaryModuleClassResolver(sourceScope)
        
        val languageVersionSettings = LanguageVersionSettingsImpl.DEFAULT
        val optionalBuiltInsModule = JvmBuiltIns(storageManager).apply { initialize(module, true) }.builtInsModule
        
        val dependencyModule = run { 
            val dependenciesContext = ContextForNewModule(
                    moduleContext,  Name.special("<dependencies of ${environment.configuration.getNotNull<String>(CommonConfigurationKeys.MODULE_NAME)}>"),
                    module.builtIns, null
            )
            
            val dependencyScope = GlobalSearchScope.notScope(sourceScope)
            val dependenciesContainer = createContainerForTopDownAnalyzerForJvm(
                    moduleContext, 
                    trace, 
                    providerFactory, 
                    dependencyScope, 
                    LookupTracker.DO_NOTHING,
                    KotlinPackagePartProvider(kotlinProject), 
                    languageVersionSettings,
                    moduleClassResolver,
                    kotlinProject)
            
            moduleClassResolver.compiledCodeResolver = dependenciesContainer.get<JavaDescriptorResolver>()
            
            dependenciesContext.setDependencies(listOfNotNull(dependenciesContext.module, optionalBuiltInsModule))
            dependenciesContext.initializeModuleContents(CompositePackageFragmentProvider(listOf(
                    moduleClassResolver.compiledCodeResolver.packageFragmentProvider,
                    dependenciesContainer.get<JvmBuiltInsPackageFragmentProvider>()
            )))
            
            dependenciesContext.module
        }
        
        val container = createContainerForTopDownAnalyzerForJvm(
                moduleContext, 
                trace, 
                providerFactory, 
                sourceScope, 
                LookupTracker.DO_NOTHING, 
                KotlinPackagePartProvider(kotlinProject), 
                languageVersionSettings, 
                moduleClassResolver, 
                kotlinProject).apply {
            initJvmBuiltInsForTopDownAnalysis(module, languageVersionSettings)
        }
        
        moduleClassResolver.sourceCodeResolver = container.get<JavaDescriptorResolver>()
        
        val additionalProviders = arrayListOf<PackageFragmentProvider>()
        additionalProviders.add(container.get<JavaDescriptorResolver>().packageFragmentProvider)
        
        PackageFragmentProviderExtension.getInstances(project).mapNotNullTo(additionalProviders) { extension ->
            extension.getPackageFragmentProvider(project, module, storageManager, trace, null)
        }
        
        module.setDependencies(ModuleDependenciesImpl(
                listOfNotNull(module, dependencyModule, optionalBuiltInsModule),
                setOf(dependencyModule)
        ))
        module.initialize(CompositePackageFragmentProvider(
                listOf(container.get<KotlinCodeAnalyzer>().packageFragmentProvider) + additionalProviders
        ))
        
        try {
            container.get<LazyTopDownAnalyzer>().analyzeDeclarations(TopDownAnalysisMode.TopLevelDeclarations, filesSet)
        } catch (e: KotlinFrontEndException) {}
        
        return AnalysisResultWithProvider(
                AnalysisResult.success(trace.bindingContext, module),
                container
        )
    }
    
    private fun getPath(jetFile: KtFile): String? = jetFile.getVirtualFile()?.getPath()
    
    private fun createModuleContext(project: Project,
                                    configuration: CompilerConfiguration,
                                    createBuiltInsFromModule: Boolean): MutableModuleContext {
        val projectContext = ProjectContext(project)
        val builtIns = JvmBuiltIns(projectContext.storageManager, !createBuiltInsFromModule)
        
        return ContextForNewModule(
                projectContext, Name.special("<${configuration.getNotNull<String>(CommonConfigurationKeys.MODULE_NAME)}>"), builtIns, null
        ).apply { 
            if (createBuiltInsFromModule) {
                builtIns.builtInsModule = module
            }
        }
        
    }
    
}